File: python/Arrays&Hashing/top_k_frequent_elements.py
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k : int) -> List[int]:
        """
        Approach:
        - Count the frequency of each element using a dictionary.
        - Sort the items of the dictionary by frequency in descending order.
        - Return the first k elements from the sorted list.

        Time Complexity: O(n log n), where n is the number of elements in nums (due to sorting).
        Space Complexity: O(n), for the frequency dictionary and the output list.
        """
        freq = {}
        for i in nums:
            if i in freq:
                freq[i] += 1
            else:
                freq[i] = 1
        sorted_items = sorted(freq.items(), key=lambda x: x[1], reverse=True)
        return [item[0] for item in sorted_items[:k]]

# -------------------
# Test Cases
# -------------------
def test_top_k_frequent_elements():
    sol = Solution()
    # Test case 1: Example from Leetcode
    nums1 = [1,1,1,2,2,3]
    assert set(sol.topKFrequent(nums1, 2)) == {1,2}
    # Test case 2: All unique
    nums2 = [1,2,3,4]
    assert set(sol.topKFrequent(nums2, 2)) == {1,2,3,4} - set(sol.topKFrequent(nums2, 2)) == set()
    # Test case 3: k = 1
    nums3 = [4,4,4,6,6,7]
    assert sol.topKFrequent(nums3, 1) == [4]
    # Test case 4: k equals number of unique elements
    nums4 = [5,6,7,8]
    assert set(sol.topKFrequent(nums4, 4)) == {5,6,7,8}
    # Test case 5: Single element
    nums5 = [42]
    assert sol.topKFrequent(nums5, 1) == [42]
    print("All test cases passed!")

if __name__ == "__main__":
    test_top_k_frequent_elements()

File: C++/Arrays&Hashing/top_k_frequent_elements.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Problem: Find the k most frequent elements in array
     * 
     * Approach:
     * 1. Count frequencies using hash map
     * 2. Convert to vector of pairs (frequency, number)
     * 3. Sort pairs by frequency in descending order
     * 4. Return first k numbers
     * 
     * Time Complexity: O(n log n)
     * - n for counting frequencies O(n)
     * - Converting to vector O(n)
     * - Sorting vector O(n log n)
     * - Extracting k elements O(k)
     * Overall: O(n log n) dominated by sort
     * 
     * Space Complexity: O(n)
     * - Hash map stores at most n elements
     * - Vector of pairs stores at most n pairs
     * - Result vector stores k elements
     */
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Count frequencies using hash map
        unordered_map<int, int> freq;
        for (int num : nums) {
            freq[num]++;
        }
        
        // Convert to vector of pairs for sorting
        vector<pair<int, int>> freqVec;
        for (const auto& entry : freq) {
            freqVec.push_back({entry.second, entry.first});
        }
        
        // Sort by frequency in descending order
        sort(freqVec.begin(), freqVec.end(), greater<pair<int, int>>());
        
        // Get top k elements
        vector<int> result;
        for (int i = 0; i < k; i++) {
            result.push_back(freqVec[i].second);
        }
        
        return result;
    }
};

/*
 * Detailed Syntax Explanation:
 * 
 * 1. Function Declaration:
 *    vector<int> topKFrequent(vector<int>& nums, int k)
 *    - Returns vector of k most frequent integers
 *    - Takes reference to input array and k value
 *    - & prevents copying of input vector
 * 
 * 2. Frequency Counting:
 *    unordered_map<int, int> freq;
 *    - Key: number from array
 *    - Value: its frequency count
 *    - O(1) average lookup/insert
 *    freq[num]++
 *    - Increment count for each number
 * 
 * 3. Pair Vector Creation:
 *    vector<pair<int, int>> freqVec;
 *    - Stores {frequency, number} pairs
 *    - First element is frequency for sorting
 *    freqVec.push_back({entry.second, entry.first})
 *    - Creates pair with frequency first
 * 
 * 4. Sorting:
 *    sort(freqVec.begin(), freqVec.end(), greater<pair<int,int>>())
 *    - STL sort with custom comparator
 *    - greater<> sorts in descending order
 *    - Compares first elements (frequencies)
 * 
 * 5. Result Collection:
 *    vector<int> result;
 *    - Stores k most frequent numbers
 *    result.push_back(freqVec[i].second)
 *    - second gives the number (not frequency)
 */
 
